autoretrieve/engine.go

```diff
@@ -0,0 +1,524 @@
+			// new advertisement with same linked list, but new metadata.
+			cidsLnk = cidlink.Link{Cid: c}
+		}
+
+		if err = e.putKeyMetadataMap(ctx, contextID, &md); err != nil {
+			return cid.Undef, fmt.Errorf("failed to write context id to metadata mapping: %s", err)
+		}
+	} else {
+		log.Info("Creating removal advertisement")
+
+		if c == cid.Undef {
+			return cid.Undef, provider.ErrContextIDNotFound
+		}
+
+		// And if we are removing it means we probably do not have the list of
+		// CIDs anymore, so we can remove the entry from the datastore.
+		err = e.deleteKeyCidMap(ctx, contextID)
+		if err != nil {
+			return cid.Undef, fmt.Errorf("failed to delete context id to entries cid mapping: %s", err)
+		}
+		err = e.deleteCidKeyMap(ctx, c)
+		if err != nil {
+			return cid.Undef, fmt.Errorf("failed to delete entries cid to context id mapping: %s", err)
+		}
+		err = e.deleteKeyMetadataMap(ctx, contextID)
+		if err != nil {
+			return cid.Undef, fmt.Errorf("failed to delete context id to metadata mapping: %s", err)
+		}
+
+		// Create an advertisement to delete content by contextID by specifying
+		// that advertisement has no entries.
+		cidsLnk = schema.NoEntries
+
+		// The advertisement still requires a valid metadata even though
+		// metadata is not used for removal.  Create a valid empty metadata.
+		md = metadata.New(metadata.Bitswap{})
+	}
+
+	mdBytes, err := md.MarshalBinary()
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	adv := schema.Advertisement{
+		Provider:  providerID,
+		Addresses: addresses,
+		Entries:   cidsLnk,
+		ContextID: contextID,
+		Metadata:  mdBytes,
+		IsRm:      isRm,
+	}
+
+	// Get the previous advertisement that was generated
+	prevAdvID, err := e.getLatestAdCid(ctx)
+	if err != nil {
+		return cid.Undef, fmt.Errorf("could not get latest advertisement: %s", err)
+	}
+
+	// Check for cid.Undef for the previous link. If this is the case, then
+	// this means there is a "cid too short" error in IPLD links serialization.
+	if prevAdvID != cid.Undef {
+		log.Info("Latest advertisement CID was undefined - no previous advertisement")
+		prev := ipld.Link(cidlink.Link{Cid: prevAdvID})
+		adv.PreviousID = &prev
+	}
+
+	// Sign the advertisement.
+	if err := adv.Sign(e.key); err != nil {
+		return cid.Undef, err
+	}
+	return e.Publish(ctx, adv)
+}
+
+func (e *AutoretrieveEngine) putKeyCidMap(ctx context.Context, contextID []byte, c cid.Cid) error {
+	// We need to store the map Key-Cid to know what CidLink to put
+	// in advertisement when we notify a removal.
+	err := e.ds.Put(ctx, datastore.NewKey(keyToCidMapPrefix+string(contextID)), c.Bytes())
+	if err != nil {
+		return err
+	}
+	// And the other way around when graphsync ios making a request,
+	// so the lister in the linksystem knows to what contextID we are referring.
+	return e.ds.Put(ctx, datastore.NewKey(cidToKeyMapPrefix+c.String()), contextID)
+}
+
+func (e *AutoretrieveEngine) getKeyCidMap(ctx context.Context, contextID []byte) (cid.Cid, error) {
```

Comment:
can we change the `key` in these functions to `contextID` - `key` is too generic also, no need for the `map` suffix. - `getContextIDCID` , `putContextIDCID` etc

Reply:
100% agree, but think it's better suited for https://github.com/filecoin-project/index-provider/

Reply:
Again, I think our implementation can be different from the reference repo